### PR TITLE
Add new funcionnality in the plug-in

### PR DESCRIPTION
**Pull request by AlexKuen**

### CODE REVIEW

File: src/sot_gepetto_viewer/command_execution.py [hunk 1]
```diff
@@ -82,8 +82,9 @@ def runRemoteCode(self, code, retValue, retry = True):
                 print(response.standarderror[:-1])
             elif response.result != "None":
                 if not retValue: return
-                res = eval(response.result)
-                # print(response.result)
+		print(response.result)
+                res = response.result
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/command_execution.py:85]: - You have a mix of tabulations and spaces.
- Is the `print` meaningful ? I think it is only for debugging purpose.
- What was wrong with the `eval` ? The result of the command should be evaluated to have the same behavior locally and remotely.

File: src/sot_gepetto_viewer/graph.py [hunk 2]
```diff
@@ -54,30 +63,95 @@ def initCmd (self):
...
+        print ("Auto Refresh Stopped")
+
+    def LaunchRefresh(self):
+        print ("Auto Refreshed Launched")
+        self.timer = QtCore.QTimer()
+        self.timer.connect(self.timer, QtCore.SIGNAL("timeout()"), self.createAllGraph)
+        self.timer.start(1000)
+    
+    def UpdateFilter (self, Filt):
+        self.filter = Filt.split()
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: The class non-static methods should follow one unique naming convention. Make them start with a small letter, unless you have a good reason not to do so.

File: src/sot_gepetto_viewer/graph.py [hunk 3]
```diff
@@ -54,30 +63,95 @@ def initCmd (self):
...
+
+    def getList (self):
+        chaine = "\n"
+        liste = dynamic_graph.entity.Entity.entityClassNameList
+        for i in liste:
+            chaine += i + "\n"
+
+        QMessageBox.information(None, "Entities",chaine,
+                        QMessageBox.Ok,
+                        QMessageBox.Ok)
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py:90]: I think this function was added for debugging and should be removed.

File: src/sot_gepetto_viewer/graph.py [hunk 4]
```diff
@@ -1,6 +1,13 @@
+from __future__ import print_function
 from PythonQt.QGraphViz import QGVScene
+from PythonQt import QtCore
+from PythonQt.QtCore import QThreadPool, QRunnable
 from PythonQt import QtGui, Qt
+from PythonQt.QtGui import QMessageBox
 from command_execution import CommandExecution
+import dynamic_graph
+from dynamic_graph.sot.core.robot_simu import RobotSimu
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: You shouldn't import anything from `dynamic_graph` module in this file. This breaks the separation between SoT and this viewer.

If there are functions which needs this (and thus do not work when using remote SoT), then you should put the import inside those functions.

File: src/sot_gepetto_viewer/graph.py [hunk 5]
```diff
@@ -54,30 +63,95 @@ def initCmd (self):
...
-                self._nodeEntity(e)
-        for e in entities:
-            if self.entityFilter is not None and not self.entityFilter.search(e):
-                continue
-            etype = self.types[e]
-            if self.typeCallbacks.has_key(etype):
-                self.typeCallbacks[etype][1] (e)
-            else:
-                self._edgeEntitySignals (e)
+        entities = eval(str_entities)
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: As mentioned earlier, this `eval` should be moved inside `self.cmd.run`.

File: src/sot_gepetto_viewer/graph.py [hunk 6]
```diff
@@ -54,30 +63,95 @@ def initCmd (self):
...
+
+        QMessageBox.information(None, "Entities",chaine,
+                        QMessageBox.Ok,
+                        QMessageBox.Ok)
+
 
     def createAllGraph (self):
-        entities = self.cmd.run ("dg.entity.Entity.entities.keys()")
+        self.blocked =""
+        self.blocked2 =""
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: Can you rename `blocked` and / or `blocked2` ?

File: src/sot_gepetto_viewer/graph.py [hunk 7]
```diff
@@ -54,30 +63,95 @@ def initCmd (self):
...
+                        if self.entityFilter is not None and not self.entityFilter.search(e):
+                            continue
+                        etype = self.cmd.run("dg.entity.Entity.entities['"+e+"'].className")
+                        self.types[e] = etype
+                        if self.typeCallbacks.has_key(etype):
+                            self.typeCallbacks[etype][0] (e)
+                        else:
+                            self._nodeEntity(e)
+                        again += 1
+            again = 0
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: Suggestion to make the code easier to read. Note that:
- you should apply it to the loop below,
- you should test it.

```suggestion
        for e in ent_list:
            for i in self.filter:
                if i != "0" and i not in e:
                    continue
                if self.entityFilter is not None and not self.entityFilter.search(e):
                    continue
                etype = self.cmd.run("dg.entity.Entity.entities['"+e+"'].className")
                self.types[e] = etype
                if self.typeCallbacks.has_key(etype):
                    self.typeCallbacks[etype][0] (e)
                else:
                    self._nodeEntity(e)
                break
```

File: src/sot_gepetto_viewer/graph.py [hunk 8]
```diff
@@ -54,30 +63,95 @@ def initCmd (self):
...
+                            self.typeCallbacks[etype][1] (e)
+                        else:
+                            self._edgeEntitySignals (e)
+                        again += 1
+            again = 0
         self.initLayout()
 
+    def getNodeInformation (self, e):
+        str_signals = self.cmd.run("[ s.name for s in dg.entity.Entity.entities['"+e+"'].signals() ]")
+        signals = eval(str_signals)
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: Again, the eval should not be here.

File: src/sot_gepetto_viewer/graph.py [hunk 9]
```diff
@@ -125,11 +199,13 @@ def _edgeEntitySOT(self, s):
         for t,n in zip(tasks, nodes):
             if not self.nodes.has_key(t):
                 self._createGraphBackwardFromEntity(t)
-            e = self.graph.addEdge (self.nodes[t], n, "error")
-            # TODO errorTimeDerivative
-            n = "sot_" + s + "/task_" + t + "/error"
-            self.edges[n] = (t, e)
-            self.edgesBack[e] = n
+            for i in self.filter:
+                if i == "0":
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: Isn't this sufficient ?
```suggestion
            if "0" in self.filter:
```

File: src/sot_gepetto_viewer/graph.py [hunk 10]
```diff
@@ -125,11 +199,13 @@ def _edgeEntitySOT(self, s):
...
+            	    e = self.graph.addEdge (self.nodes[t], n, "error")
+                    # TODO errorTimeDerivative
```
Comment (jmirabel) [inline src/sot_gepetto_viewer/graph.py]: You have a weird indentation here.

File: README.md [hunk 11]
```diff
@@ -1,3 +1,77 @@
...
+Confiurate gepetto-gui
+ 1. Open a terminal and do "gepetto-gui"
+ Gepetto-gui will now open.
+ 2. click on "Plugins", normally you will only see OmniORB Server detected and loaded.
+ 3. Click on "Find all lugins", PyPCustomPlot and PyQGV will be detected, load them if they are not already loaded (right click then click load)
+ 4. click on save
+
+gepetto-gui is now configure.
+
+If you want to launch the new plugin, you will need to write "gepetto-gui --load-pyplugin sot_gepetto_viewer.sot_plugin" in the terminal
```
Comment (jmirabel) [inline README.md:77]: Before I go in deeper details for this file, is this till up-to-date ?

**Review by jmirabel — CHANGES_REQUESTED** [review]
Hi,

thanks for your work. I made some comments in order to improve code clarity (mostly) and to keep the ability to use local and remote SoT.

I did not comment everything. Many of the comments I made apply to what I didn't review.

File: src/sot_gepetto_viewer/command_execution.py (hunk 1 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/command_execution.py:85]: There was a tabulation between spaces.
The print was only for debuggingso I deleted it
The plug-in won't work if the eval() was still here

File: src/sot_gepetto_viewer/graph.py (hunk 2 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/graph.py]: There was no color on Ubuntu for Filt so i hadn't seen it, i renamed it

File: src/sot_gepetto_viewer/graph.py (hunk 3 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/graph.py:90]: I use it to show the list of used entities, QMessageBox is not just for debuggin.

File: src/sot_gepetto_viewer/graph.py (hunk 4 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/graph.py]: Moved them in the function where they are needed

File: src/sot_gepetto_viewer/graph.py (hunk 7 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/graph.py]: Good point.
I tested your suggestion, it worked, so i used it to change the loops for the creation of entities and signals.

File: src/sot_gepetto_viewer/graph.py (hunk 10 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/graph.py]: One tabulation at the start of the line caused the indentation problem

File: README.md (hunk 11 above)
Comment (AlexKuen) [reply README.md:77]: Yes, it is up to date. I used it 2 weeks ago to install the plug-in on a new VM

File: src/sot_gepetto_viewer/graph.py (hunk 9 above)
Comment (AlexKuen) [reply src/sot_gepetto_viewer/graph.py]: It's sufficient, i changed it for that

File: src/sot_gepetto_viewer/graph.py (hunk 2 above)
Comment (jmirabel) [reply src/sot_gepetto_viewer/graph.py]: I refered to the class method too.

**Pull request closed without merging**